zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/QuorumCnxManager.java

```diff
@@ -228,28 +240,33 @@ static public InitialMessage parse(Long protocolVersion, DataInputStream din)
                         num_read, remaining, sid);
             }
 
-            String addr = new String(b);
-            String[] host_port;
-            try {
-                host_port = ConfigUtils.getHostAndPort(addr);
-            } catch (ConfigException e) {
-                throw new InitialMessageException("Badly formed address: %s", addr);
-            }
+            String[] addressStrings = new String(b).split(",");
```

Comment:
As discussed offline, we probably need 2 things here:
- Increment PROTOCOL_VERSION, because the patch introduces new InitialMessage format,
- Use JSON format and add Jackson JSON serialiser to enable forward compatibility for future changes.

Reply:
referring to our second offline discussion:
- I incremented the PROTOCOL_VERSION
- did not do the JSON format modification, as it makes the future code complicated to always ensure both forward and backward compatibility of the election protocol during the rolling upgrades. The current solution (simply failing if the protocol versions mismatch) is more simple and still working just fine: as the servers are restarted one-by-one, the nodes with the old protocol version and the nodes with the new protocol version will form two partitions, but any given time only one partition will have the quorum.